File: output/HeadServer.js
require('./mscorlib.js');require('./MongoDBLibrary.js');require('./CommonServerLibraries.js');require('./CommonLibraries.js');require('./CommonShuffleLibrary.js');require('./Models.js');
(function() {
	////////////////////////////////////////////////////////////////////////////////
	// HeadServer.HeadServer
	var $HeadServer_HeadServer = function() {
		this.$__dirname = CommonLibraries.ExtensionMethods.HARDLOCATION;
		this.$fs = require('fs');
		this.$gateways = [];
		this.$indexForSites = [];
		this.$indexPageData = null;
		this.$oldGateways = [];
		this.$oldIndex = [];
		this.$pubsub = null;
		this.$qManager = null;
		this.$siteIndex = 0;
		var name = 'Head1';
		CommonServerLibraries.Logger.start(name);
		this.$qManager = new CommonShuffleLibrary.QueueManager(name, new CommonShuffleLibrary.QueueManagerOptions([new CommonShuffleLibrary.QueueWatcher('HeadServer', null), new CommonShuffleLibrary.QueueWatcher(name, null)], ['GatewayServer']));
		this.$fs.readFile(this.$__dirname + '/index.html', 'ascii', ss.mkdel(this, this.ready));
		this.$pubsub = new CommonShuffleLibrary.PubSub(ss.mkdel(this, function() {
			this.$pubsub.subscribe(String).call(this.$pubsub, 'PUBSUB.GatewayServers', ss.mkdel(this, function(message) {
				ss.add(this.$indexForSites, ss.replaceAllString(this.$indexPageData, '{{gateway}}', message));
				ss.add(this.$gateways, message);
			}));
		}));
		require('http').createServer(ss.mkdel(this, this.$handlerWS)).listen(8844);
		setInterval(ss.mkdel(this, this.$pollGateways), 5000);
		this.$pollGateways();
	};
	$HeadServer_HeadServer.prototype = {
		$pollGateways: function() {
			this.$pubsub.publish('PUBSUB.GatewayServers.Ping', '');
			if (this.$indexForSites.length > 0) {
				this.$oldIndex = this.$indexForSites;
			}
			if (this.$gateways.length > 0) {
				this.$oldGateways = this.$gateways;
			}
			this.$indexForSites = [];
			this.$gateways = [];
			this.$siteIndex = 0;
		},
		$handlerWS: function(request, response) {
			if (this.$oldGateways.length > 0) {
				var inj = this.$siteIndex++ % this.$oldIndex.length;
				response.end(this.$oldGateways[inj]);
				return;
			}
			response.end();
		},
		$handler: function(request, response) {
			var dict = {};
			dict['Content-Type'] = 'text/html';
			if (this.$oldIndex.length > 0) {
				response.writeHead(200, dict);
				var inj = this.$siteIndex++ % this.$oldIndex.length;
				response.end(this.$oldIndex[inj]);
			}
			else {
				response.writeHead(200, dict);
				response.end();
			}
		},
		ready: function(error, content) {
			this.$indexPageData = content.toString();
			this.$indexPageData = content.toString();
			require('http').createServer(ss.mkdel(this, this.$handler)).listen(80);
		}
	};
	$HeadServer_HeadServer.main = function() {
		try {
			new $HeadServer_HeadServer();
		}
		catch ($t1) {
			var exc = ss.Exception.wrap($t1);
			CommonServerLibraries.Logger.log('CRITICAL FAILURE: ' + CommonLibraries.ExtensionMethods.goodMessage(exc), 0);
		}
	};
	ss.registerClass(global, 'HeadServer.HeadServer', $HeadServer_HeadServer);
	$HeadServer_HeadServer.main();
})();

File: Servers/HeadServer/HeadServer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using CommonLibraries;
using CommonShuffleLibrary;
using NodeJSLibrary;
namespace HeadServer
{
    public class HeadServer
    {
        private string __dirname = "/usr/local/src/new";
        private FS fs = Global.Require<FS>("fs");
        private List<string> gateways = new List<string>();
        private List<string> indexForSites = new List<string>();
        private string indexPageData;
        private List<string> oldGateways = new List<string>();
        private List<string> oldIndex = new List<string>();
        private PubSub pubsub;
        private QueueManager qManager;
        private int siteIndex;

        public HeadServer()
        {
            qManager = new QueueManager("Head1",
                                        new QueueManagerOptions(new[] {
                                                                              new QueueWatcher("HeadServer", null),
                                                                              new QueueWatcher("Head1", null),
                                                                      },
                                                                new[] {
                                                                              "GatewayServer"
                                                                      }));

            fs.ReadFile(__dirname + "/index.html", "ascii", ready);

            pubsub = new PubSub(() => pubsub.Subscribe<string>("PUBSUB.GatewayServers",
                                                               message => {
                                                                   indexForSites.Add(indexPageData.Replace("{{gateway}}", message));
                                                                   gateways.Add(message);
                                                               }));

            Global.Require<Http>("http").CreateServer(handlerWS).Listen(8844);

            Global.SetInterval(pollGateways, 1000);
            pollGateways();
        }

        private void pollGateways()
        {
            pubsub.Publish("PUBSUB.GatewayServers.Ping", "");

            if (indexForSites.Count > 0)
                oldIndex = indexForSites;
            if (gateways.Count > 0)
                oldGateways = gateways;
            indexForSites = new List<string>();
            gateways = new List<string>();
            siteIndex = 0;
        }

        private void handlerWS(HttpRequest request, HttpResponse response)
        {
            if (oldGateways.Count > 0) {
                var inj = ( siteIndex++ ) % oldIndex.Count;
                response.End(oldGateways[inj]);
                return;
            }
            response.End();
        }

        private void handler(HttpRequest request, HttpResponse response)
        {
            var dict = new JsDictionary();
            dict["Content-Type"] = "text/html";
            if (oldIndex.Count > 0) {
                response.WriteHead(200, dict);
                var inj = ( siteIndex++ ) % oldIndex.Count;
                response.End(oldIndex[inj]);
            } else {
                response.WriteHead(200, dict);
                response.End();
            }
        }

        public void ready(FileSystemError error, object content)
        {
            indexPageData = content.ToString();
            Global.Require<Http>("http").CreateServer(handler).Listen(80);
        }

        public static void Main()
        {
            try {
                new HeadServer();
            } catch (Exception exc) {
                Console.Log("CRITICAL FAILURE: " + exc.GoodMessage());
            }
        }
    }
}